Force hook on first run

shanesmith opened the pull request:
There are scenarios where the shadowenv's exported environment is carried forward to a new shell as expected but is modified by the new shell's initialization. In this case shadowenv would not update the environment since the working directory hasn't changed from the previous shell.

The example below shows what happens when the user defines a default ruby in their bashrc.

```
$ cat ~/.bashrc
chruby 2.7.0

$ dev cd web
activated shadowenv (node:v10.18.0, ruby:2.6.5)

$ ruby --version
ruby 2.6.5

$ bash -l

$ ruby --version
ruby 2.7.0
```

Some affected scenarios that I know of include:

- Starting a new shell from a project directory (as exampled above)
- Starting tmux from a project directory
- Starting an editor (vscode, vim...) from a project directory and using its built-in terminal

To address the issue this PR updates the shell hook functions to set an _unexported_ variable named `$__shadowenv_first_run`. The hook will look for the variable and if it's not defined, which it wont be in a new shell because it is unexported, it would mean that it is the first run of the hook and it will pass the new `--force` flag to the `shadowenv hook` command. The `--force` flag instructs shadowenv always apply the new environment regardless of whether the working directory changed.

Fixes https://github.com/Shopify/vscode-shadowenv/issues/4

Also seen on [discourse](https://discourse.shopify.io/t/shadowenv-not-being-activated-unless-directory-is-explicitly-cdd-into/8478/3)

lugray commented on src/hook.rs:
```diff
@@ -79,7 +81,7 @@ pub fn load_env(
+        (Some(a), Some(t)) if a.hash == t.hash()? && !force => {
             return Ok(None);
         }
         (_, _) => (),
```
It looks to me like when `__shadowenv_data`, which contains the `active`, was unexported, the new shell scenario would have led to `(None, Some(t))`, caught in this default case.  This PR returns us to that case.  I do wonder though, if it also reintroduces what exporting `__shadowenv_data` in #67 fixed. @burke can you comment on the nature of the tmux problems so that @shanesmith can check whether they come back with this change?

burke replied:
Yeah, the basic issue was:

1. User starts a shell, `dev cd shopify`, activates core env
2. User starts tmux. Tmux inherits core env, but not undo data
3. User moves to other project. Because the core env is applied, but the undo data is gone, shadowenv just merges this project's env into core's env.

shanesmith commented on src/cli.rs:
```diff
@@ -33,6 +33,11 @@ pub fn app() -> App<'static, 'static> {
+                .arg(
+                    Arg::with_name("force")
+                        .long("force")
+                        .help("USE THE FORCE"),
```
hehe yeah I would have changed this before merging of course ^^;

burke commented on sh/shadowenv.zsh.in:
```diff
@@ -3,7 +3,11 @@ __shadowenv_hook() {
   if [[ "$1" == "zsh-preexec" ]]; then
     flags=(--silent)
   fi
+  if [[ ! -v __shadowenv_first_run ]]; then
```
there'll be a fish one to change too I think

burke commented on sh/shadowenv.bash.in:
```diff
@@ -3,9 +3,13 @@ __shadowenv_hook() {
   if [[ "$1" == "preexec" ]]; then
     flags+=(--silent)
   fi
```
I think it would be nicer if we did this the other way around, defining the variable outside of the function, and unsetting it after the first run. Would that work?

shanesmith replied:
I liked how keeping it all in the function made it a little more self-contained and wouldn't depend on something being set elsewhere, but it has made it a little awkward to name the variable... is that mainly why you'd think it would be nicer if defined outside?

no strong feelings at the end of the day though

lugray replied:
The reason I prefer what @burke suggests is that it leaves the environment unpolluted by an environment variable for most of the time.  It's something we strive for in the tools we write.

shanesmith replied:
that's fair, i'll update

shanesmith commented:
updated with the following

- updated fish shell script
- shell scripts now force when `__shadowenv_force_run` is set
- updated help message for the `--force` flag
- `shadowenv exec` now loads the shadow env with `force: true` (seems like the right thing to do, let me know otherwise)

shanesmith commented:
> Looks good, provided you tested [#72 (comment)](https://github.com/Shopify/shadowenv/pull/72#discussion_r431234030) and all's well.

yup, tested and everything seems fine =)
